packages/server/src/user/user.service.ts

```diff
@@ -115,34 +143,53 @@ export class UserService {
   /**
    * Generate reset token for a user and update it in the database.
    *
-   * @param params object includes: `project_id` and `email`
-   * @param reset_code a string of reset code in plain text
+   * @param params object includes: `project_id` and `email` as string
+   * @param reset_code_plain a string of reset code in plain text
    */
-  async setResetToken(params: { project_id; email }, reset_code: string) {
-    // TODO: Replace set reset token for temperary user
-    for (let idx = 0; idx < this.users.length; idx++) {
-      const user = this.users[idx];
-      if (params.project_id === user.project_id && params.email === user.email) {
-        user.reset_code = reset_code;
-        break;
+  async setResetToken(params: { project_id; email }, reset_code_plain: string): Promise<void> {
+    const reset_code_hash = await argon2.hash(reset_code_plain);
+
+    const valid_time = 60 * 60 * 1000; // valid time for reset code in millisecond
+    const user = await this.prisma.uSER.update({
+      where: {
+        proj_email: params
+      },
+      data: {
+        reset_code: reset_code_hash,
+        reset_code_expires_at: new Date(Date.now() + valid_time)
       }
-    }
+    });
   }
 
   /**
    * Update user password after verifying user's reset_code against the database
    *
-   * @param params object includes: `project_id`, `email` and new `password` in plain text
-   * @param reset_code a string of reset code in plain text
+   * @param params object includes: `project_id` and `email` as string
+   * @param pwd_plain password in plain text
+   * @param reset_code_plain a string of reset code in plain text
    */
-  async updateUserPassword(params: { project_id; email; password }, reset_code: string) {
-    // TODO: Replace update password for temperary user
-    for (let idx = 0; idx < this.users.length; idx++) {
-      const user = this.users[idx];
-      if (params.project_id === user.project_id && params.email === user.email && reset_code === user.reset_code) {
-        user.reset_code = null;
-        user.password = params.password;
+  async updateUserPassword(params: { project_id; email }, pwd_plain: string, reset_code_plain: string): Promise<void> {
+    // check if reset code matches
+    const user = await this.prisma.uSER.findUniqueOrThrow({
+      where: {
+        proj_email: params
       }
+    });
+
+    const now = new Date(Date.now());
+    if ((await argon2.verify(user.reset_code, reset_code_plain)) && now < user.reset_code_expires_at) {
```

Comment:
https://date-fns.org/v2.29.3/docs/isFuture

```suggestion
    if ((await argon2.verify(user.reset_code, reset_code_plain)) && isFutrue(user.reset_code_expires_at) {
```